utcp.c

```
@@ -509,15 +627,157 @@ static void retransmit(struct utcp_connection *c) {
+ * - the SACK entry starts before and ends after ^, so we have to
+ *   change both its offset and size.
+ * - the SACK entry is completely before ^, in that case delete it.
+ */
+static void sack_consume(struct utcp_connection *c, size_t len) {
+	debug("sack_consume %lu\n", (unsigned long)len);
+
+	buffer_get(&c->rcvbuf, NULL, len);
+
+	for(int i = 0; i < NSACKS && c->sacks[i].len; ) {
+		if(len < c->sacks[i].offset) {
+			c->sacks[i].offset -= len;
+			i++;
+		} else if(len < c->sacks[i].offset + c->sacks[i].len) {
+			c->sacks[i].len -= len - c->sacks[i].offset;
+			c->sacks[i].offset = 0;
+			i++;
+		} else {
+			if(i < NSACKS - 1) {
+				// move remaining sacks one ahead
+				memmove(&c->sacks[i], &c->sacks[i + 1], ((NSACKS - 1) - i) * sizeof c->sacks[i]);
+				// clean last sack len
+				c->sacks[NSACKS - 1].len = 0;
+			} else {
+				c->sacks[i].len = 0;
+				break;
+			}
+		}
+	}
+
+	for(int i = 0; i < NSACKS && c->sacks[i].len; i++)
+		debug("SACK[%d] offset %u len %u\n", i, c->sacks[i].offset, c->sacks[i].len);
+}
+
+static void handle_out_of_order(struct utcp_connection *c, uint32_t offset, const void *data, size_t len) {
+	debug("out of order packet, offset %u\n", offset);
+
+	// drop packets that are ahead of max buffer size
+	if(offset >= c->rcvbuf.maxsize) {
+		debug("warning: packet offset %u ahead of max buffer size %u\n", offset, c->rcvbuf.maxsize);
+		return;
+	}
+
+	// Packet loss or reordering occured. Store the data in the buffer.
+	ssize_t rxd = buffer_put_at(&c->rcvbuf, offset, data, len);
+
+	// Make note of where we put it.
+	for(int i = 0; i < NSACKS; i++) {
+		if(!c->sacks[i].len) { // nothing to merge, add new entry
+			debug("New SACK entry %d\n", i);
+			c->sacks[i].offset = offset;
+			c->sacks[i].len = rxd;
+			break;
+		} else if(offset < c->sacks[i].offset) {
+			if(offset + rxd < c->sacks[i].offset) { // insert before
+				if(!c->sacks[NSACKS - 1].len) { // only if room left
```

So if no room is left the received packet is silently dropped? How about at least a debug output?